scripts/lsp_types.lua

```diff
@@ -18,19 +19,90 @@ local function tofile(fname, text)
   end
 end
 
-function M.gen(opt)
-  if vim.uv.fs_stat('./lsp.json') then
-    vim.fn.delete('./lsp.json')
+local function read_json(opt)
+  local uri = 'https://raw.githubusercontent.com/microsoft/language-server-protocol/gh-pages/_specifications/lsp/'
+    .. opt.version
+    .. '/metaModel/metaModel.json'
+
+  local res = vim
+    .system({
+      'curl',
+      uri,
+      '-o',
+      '-',
+    })
+    :wait()
+
+  if res.code ~= 0 then
+    print(res.stderr)
+    return
   end
-  vim.fn.system({
-    'curl',
-    'https://raw.githubusercontent.com/microsoft/lsprotocol/' .. opt.ref .. '/generator/lsp.json',
-    '-o',
-    './lsp.json',
+  return vim.json.decode(res.stdout)
+end
+
+local function gen_methods(protocol)
+  local output =
+    { '--- LSP methods. Generated by lsp_types.lua, keep at end of file.', 'protocol.Methods = {' }
+  local indent = (' '):rep(2)
+
+  vim.iter(protocol.requests):map(function(item)
+    if item.method then
+      local name = item.method:gsub('/', '_', 3)
+      if item.documentation then
+        local document = vim.split(item.documentation, '\n?\n', { trimempty = true })
+        vim.iter(document):map(function(docstring)
+          output[#output + 1] = indent .. '--- ' .. docstring
+        end)
+      end
+      output[#output + 1] = indent .. name .. " = '" .. item.method .. "',"
+    end
+  end)
+  output[#output + 1] = '}'
+  output = vim.list_extend(
+    output,
+    vim.split(
+      [[
+local function freeze(t)
+  return setmetatable({}, {
+    __index = t,
+    __newindex = function()
+      error('cannot modify immutable table')
+    end,
   })
-  local protocol = vim.fn.json_decode(vim.fn.readfile('./lsp.json'))
-  vim.fn.delete('./lsp.json')
-  protocol = protocol or {}
+end
+protocol.Methods = freeze(protocol.Methods)
+
+return protocol
+]],
+      '\n',
+      { trimempty = true }
+    )
+  )
+
+  local fname = './runtime/lua/vim/lsp/protocol.lua'
+  local bufnr = vim.fn.bufadd(fname)
+  vim.fn.bufload(bufnr)
+  vim.api.nvim_set_current_buf(bufnr)
+  local lines = vim.api.nvim_buf_get_lines(bufnr, 0, -1, false)
+  local index = vim.iter(ipairs(lines)):find(function(key, item)
+    return item:find('%-%-%- Lsp methods') and key or nil
+  end)
+  index = index and index - 1 or vim.api.nvim_buf_line_count(bufnr) - 1
+  vim.api.nvim_buf_set_lines(bufnr, index, -1, true, output)
+  vim.cmd.write()
+end
+
+function M.gen(opt)
+  local protocol = read_json(opt)
+  if not protocol then
+    return
+  end
+
+  if opt.version then
+    gen_methods(protocol)
```

Comment:
why is `opt.version` the field that decides if gen_methods is called? seems like it should be named `opt.methods`

Reply:
forget update new change. now the command use `--version` mean lsp version instead before `ref` (it's wried). when lsp release 3.19 we can use `gen --version 3.19 --methods --build ` to sync new methods and lsp types.